cors/dsl/cors.go

```diff
@@ -0,0 +1,182 @@
+//                Headers("X-Shared-Secret")           // One or more authorized headers, use "*" to authorize all
+//                Methods("GET", "POST")               // One or more authorized HTTP methods
+//                Expose("X-Time")                     // One or more headers exposed to clients
+//                MaxAge(600)                          // How long to cache a preflight request response
+//                Credentials()                        // Sets Access-Control-Allow-Credentials header
+//        })
+//		})
+//
+//		var _ = Service("calculator", func() {
+//        Origin("/(api|swagger)[.]goa[.]design/") // Define CORS policy with a regular expression
+//
+//        Method("add", func() {
+//            Description("Add two operands")
+//
+//            // Override default service security requirements. Require
+//            // both basic auth and OAuth2 access token with "api:write"
+//            // scope.
+//            Security(BasicAuth, "oauth2", func() {
+//                Scope("api:write")
+//            })
+//
+//            Payload(Operands)
+//            Error(ErrBadRequest, ErrorResult)
+//        })
+//		})
+//
+func Origin(origin string, args ...interface{}) {
+	o := &design.OriginExpr{Origin: origin}
+	if strings.HasPrefix(origin, "/") && strings.HasSuffix(origin, "/") {
+		o.Regexp = true
+		o.Origin = strings.Trim(origin, "/")
+	}
+
+	var dsl func()
+	{
+		if d, ok := args[len(args)-1].(func()); ok {
+			args = args[:len(args)-1]
+			dsl = d
+		}
+	}
+	if dsl != nil {
+		if !eval.Execute(dsl, o) {
+			return
+		}
+	}
+
+	current := eval.Current()
+	switch current.(type) {
+	case *goadesign.APIExpr:
+		design.Root.APIOrigins[origin] = o
+	case *goadesign.ServiceExpr:
+		design.Root.ServiceOrigins[origin] = o
+	default:
+		eval.IncompatibleDSL()
+		return
+	}
+	o.Parent = current
+}
+
+// Methods sets the origin allowed methods.
+//
+// Methods must be used in an Origin expression.
+//
+// Example:
+//
+// Origin("http://swagger.goa.design", func() {
+//				 Methods("GET", "POST")								// One or more authorized HTTP methods
+// })
+//
+func Methods(vals ...string) {
+	switch o := eval.Current().(type) {
+	case *design.OriginExpr:
+		o.Methods = append(o.Methods, vals...)
+	default:
+		eval.IncompatibleDSL()
+	}
+}
+
+// Expose sets the origin exposed headers.
+//
+// Expose must appear in an Origin expression.
+//
+// Example:
+//
+// Origin("http://swagger.goa.design", func() {
+//         Expose("X-Time")                     // One or more headers exposed to clients
+// })
+//
+func Expose(vals ...string) {
+	switch o := eval.Current().(type) {
+	case *design.OriginExpr:
+		o.Exposed = append(o.Exposed, vals...)
+	default:
+		eval.IncompatibleDSL()
+	}
+}
+
+// Headers sets the authorized headers. "*" authorizes all headers.
+//
+// Headers must be used in an Origin expression.
+//
+// Example:
+//
+// Origin("http://swagger.goa.design", func() {
```

Comment:
Fix indenting (code example should be indented)
Applies to all comments below.

Reply:
Done